pkg/snapshot/copier/copier.go

```diff
@@ -138,17 +142,69 @@ func (c *Copier) copyBackups() error {
 		destSnapshotsMap[snapshot.SnapName] = snapshot
 	}
 
+	// find snapshots missing in destination
+	var snapshotsToCopy brtypes.SnapList
 	for _, snapshot := range sourceSnapshot {
-		// Copy all the snapshots (if not already copied)
 		if _, ok := destSnapshotsMap[snapshot.SnapName]; !ok {
-			c.logger.Infof("Copying %s snapshot %s...", snapshot.Kind, snapshot.SnapName)
-			if err := c.copySnapshot(snapshot); err != nil {
-				return err
-			}
+			snapshotsToCopy = append(snapshotsToCopy, snapshot)
 		} else {
 			c.logger.Infof("Skipping %s snapshot %s as it already exists", snapshot.Kind, snapshot.SnapName)
 		}
 	}
+
+	if len(snapshotsToCopy) == 0 {
+		return nil
+	}
+
+	// copy all missing snapshots with configured concurrency
+	var (
+		wg     sync.WaitGroup
+		queue  = make(chan *brtypes.Snapshot, c.maxParallelCopyOperations)
+		errors = make(chan error)
+	)
+
+	// Enqueue all work items.
+	// This needs to happen in a goroutine, as the queue's buffer might not suffice to hold all work items.
+	// Close the queue and the errors channel once all snapshots have been copied.
+	go func() {
+		defer close(errors)
+		defer close(queue)
+
+		for _, snapshot := range snapshotsToCopy {
+			wg.Add(1)
+			queue <- snapshot
+		}
+
+		wg.Wait()
+	}()
+
+	// Spawn configured amount of workers for copying snapshots.
+	// Each worker takes one item at a time from the queue and exits once the queue is closed.
+	for i := 0; i < c.maxParallelCopyOperations; i++ {
+		go func() {
+			for snapshot := range queue {
+				c.logger.Infof("Copying %s snapshot %s...", snapshot.Kind, snapshot.SnapName)
+				if err := c.copySnapshot(snapshot); err != nil {
+					errors <- err
+					return
+				}
+
+				c.logger.Infof("Successfully copied %s snapshot %s...", snapshot.Kind, snapshot.SnapName)
+				wg.Done()
```

Comment:
@timebertt I just realised that `wg.Done()` will not be called when `c.copySnapshot(snapshot)` returns error, since the function returns after pushing `err` to `errors` channel. This can cause the code to hang since if even one snapshot fails to be copied, `wg.Wait()` will not allow the channels to be closed.

Can you please move this `wg.Done()` to the beginning of the goroutine, as `defer wg.Done()`?

Reply:
Nice catch! Addressed in a new commit :)